### PR TITLE
Allow overriding of host, port, protocol nsdr url path for URL building

### PR DESCRIPTION
**Pull request by pitbulk**
A developer can use setSelfProtocol, setSelfHost, setSelfPort and getBaseURLPath to define a specific value to be returned by isHTTPS, getSelfHost, getSelfPort and getBaseURLPath. And define a setBasePath to be used on the getSelfURL and getSelfRoutedURLNoQuery to replace the data extracted from $_SERVER["REQUEST_URI"].

At the settings the developer will be able to set a 'baseurl' parameter that automatically will use setBaseURL to set values for setSelfProtocol, setSelfHost, setSelfPort and setBaseURLPath.

### CODE REVIEW

**Comment by pitbulk** [comment]
@dhensby, sorry for the delay, can you review this PR?

File: lib/Saml2/Utils.php [hunk 1]
```diff
@@ -297,6 +303,41 @@ public static function redirect($url, $parameters = array(), $stay = false)
...
 
     /**
+     * @param $baseurl string The base url to be used when constructing URLs
+     */
+    public static function setBaseURL($baseurl)
+    {
+        if (!empty($baseurl)) {
+            $baseurlpath = '/';
+            if (preg_match('#^https?:\/\/([^\/]*)\/?(.*)#i', $baseurl, $matches)) {
+                if (strpos($baseurl, 'https://') == False) {
```
Comment (dhensby) [inline lib/Saml2/Utils.php]: this needs to be type equality (`===`) and you should use `false` not `False`

File: lib/Saml2/Utils.php [hunk 2]
```diff
@@ -348,6 +389,26 @@ public static function setSelfHost($host)
     }
 
     /**
+     * @param $baseurlpath string The baseurl path to use when constructing URLs
+     */
+    public static function setBaseURLPath($baseurlpath)
+    {
+        if (!isset($baseurlpath) || empty($baseurlpath)) {
```
Comment (dhensby) [inline lib/Saml2/Utils.php]: you can just used `empty` as `isset` will only return false if `$baseurlpath` is explicitly `null`, but `empty` will return true for `null` anyway.

File: lib/Saml2/Utils.php [hunk 3]
```diff
@@ -348,6 +389,26 @@ public static function setSelfHost($host)
...
+            $baseurlpath = '/';
+        }
+
+        self::$_baseurlpath = '/'. ltrim(rtrim($baseurlpath, '/') . '/', '/');
```
Comment (dhensby) [inline lib/Saml2/Utils.php]: Could you not do `'/'.trim($baseurlpath, '/') . '/';`

File: lib/Saml2/Utils.php [hunk 4]
```diff
@@ -464,9 +525,19 @@ public static function isHTTPS()
      */
     public static function getSelfURLNoQuery()
     {
-
         $selfURLhost = self::getSelfURLhost();
-        $selfURLNoQuery = $selfURLhost . $_SERVER['SCRIPT_NAME'];
+
+        if (!empty(self::getBaseURLPath())) {
```
Comment (dhensby) [inline lib/Saml2/Utils.php]: you can't use `empty` with method in php < 5.5 so you need to assign a var first as the module supports 5.3+

File: lib/Saml2/Utils.php [hunk 5]
```diff
@@ -493,6 +564,17 @@ public static function getSelfRoutedURLNoQuery()
             }
         }
 
+        if (!empty(self::getBaseURLPath())) {
```
Comment (dhensby) [inline lib/Saml2/Utils.php]: This code is repeated and could be pulled into a protected method and re-used

File: lib/Saml2/Utils.php (hunk 3 above)
Comment (pitbulk) [inline lib/Saml2/Utils.php]: I wanted to assure that all start and end with a /. What the reason for avoiding that?

File: lib/Saml2/Utils.php (hunk 4 above)
Comment (pitbulk) [inline lib/Saml2/Utils.php]: Sure, I noticed that on travis

File: lib/Saml2/Utils.php (hunk 2 above)
Comment (pitbulk) [inline lib/Saml2/Utils.php]: Ok

File: lib/Saml2/Utils.php (hunk 1 above)
Comment (pitbulk) [inline lib/Saml2/Utils.php]: Sure

**Comment by pitbulk** [comment]
Thanks, I will fix that.

File: lib/Saml2/Utils.php (hunk 3 above)
Comment (dhensby) [inline lib/Saml2/Utils.php]: Won't my suggestion have the same effect? Trim `/` off both ends then spend it to both?

Comment (pitbulk) [inline lib/Saml2/Utils.php]: haha, yes, sorry, yesterday was so late when I reviewed that

Comment (pitbulk) [inline lib/Saml2/Utils.php]: Ohh, when $baseurlpath ==> '/'

'/'.trim($baseurlpath, '/') . '/';   returns //
but '/'. ltrim(rtrim($baseurlpath, '/') . '/', '/');      returns /

Comment (dhensby) [inline lib/Saml2/Utils.php]: Ah, good point and that solution is more readable too, so cool.

**Comment by pitbulk** [comment]
any scenario we are not covering? Are we now ready to merge?

**Comment by patricknelson** [comment]
For the life of me, I can't seem to really figure out why on earth why `Util::getSelfRoutedURLNoQuery()` was setup to call `::buildWithBaseURLPath()` [here](https://github.com/onelogin/php-saml/pull/175/files#diff-5ba8e562350688b129c0428461c153c6R530). 

Also, why is it that `::getSelfRoutedURLNoQuery()` strips out the entire request URI except for the last URL segment? This causes URL's like: `https://domain/saml/acs` to be converted to `https://domain/acs` (i.e. request URI becomes just that last segment) simply because your base path is `/` and thus you end up with an error resulting during response validation:

> A valid SubjectConfirmation was not found on this Response

This is because the `$currentURL` ([from here](https://github.com/onelogin/php-saml/blob/master/lib/Saml2/Response.php#L168)) now contains `https://domain/acs` even though the `Recipient` in `SubjectConfirmationData` contains the full URL and fails the check on this line: https://github.com/onelogin/php-saml/blob/master/lib/Saml2/Response.php#L306

It seems as if this is missing a thorough explanation for the thought process behind changing the request URI like this when generating the current URL. What am I missing here?

**Comment by patricknelson** [comment]
p.s. I first mentioned this here on this comment and tracked it down to this issue. https://github.com/silverstripe/silverstripe-activedirectory/issues/104#issuecomment-323155562

**Pull request merged**